proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,754 @@
+    steps. So they are all defaulted to be matchable, at which point
+    the equality can be decomposed, and we learn that
+    ``(b :: Type -> @M Type) ~ (Id :: Type -> @U Type)``. This way, ``nested`` cannot be
+    called, because no such ``b`` exists.
+
+    Instead, we could do something more clever by defaulting
+    matchabilities in dependency order, but it's not obvious if this
+    additional complexity would be worth it.
+
+3.  We could make different choices for the syntax, regarding how we annotate
+    arrows with matchabilities or the particular names around the
+    ``Matchability`` type.
+
+4.  Data constructors could be considered to have matchable types. This would
+    make promotion more unified, as promoted constructors have matchable kinds.
+    This is quite appealing, but doing so would require additional engineering
+    effort. Either we would need to introduce matchability-polymorphic
+    term-level functions, or, perhaps more realistically, eta-expand all data
+    constructor applications to demote them to unmatchable.
+
+5.  When a kind signature is given, we make the choice of not generalising the
+    matchabilities, which differs from the treatment of kind variables. Consider
+    the following program ::
+
+      type A :: Proxy a -> Type
+      type family A
+
+    The inferred kind is ``A :: forall {k} (a :: k). Proxy a -> @M Type``, so the
+    kind of the type variable ``a`` did get generalised, but the matchability of
+    the arrow didn't (note that ``A`` is nullary). An alternative option would be
+    to simply generalise these matchability variables too, and arrive at the more
+    general ``A :: forall {k} {m} (a :: k). Proxy a -> @m Type`` kind.
+
+    But we don't do this, because doing so would result in counterintuitive
+    behaviour in many common cases, in particular, type variables introduced in
+    this way could block type family reduction. Consider the following examples ::
+
+      type B :: Type -> Type
+      type family B where
+        B = Maybe
+
+      type C :: (Type -> Type) -> Type -> Type
+      type family C f where
+        C f = f
+
+   If we infer ``B :: forall {m}. Type -> @m Type``, then ``:kind! B`` is stuck! This is
+   because type variables have computational relevance in type family reduction. In other
+   words, ``B`` becomes a matchability-indexed type family, which is likely not what the user
+   intended.
+
+   Similarly, the generalised kind of ``C`` would be
+   ``C :: forall {m} {n}. (Type -> @m Type) -> (Type -> @n Type)``, then ``:kind! C Maybe`` is stuck,
+   and so is ``:kind! C Id``.
+
+   The treatment in generalisation of matchability variables is thus different
+   from ordinary kind variables. In fact, the way kind variables are treated can also lead to unintuitive
+   behaviour ::
+
+     type T :: Proxy a
+     type family T where
+       T = 'Proxy
+
+   ``:kind! T`` is stuck, because ``T`` is indexed in the kind of ``a``, which
```

Comment:
I feel as though this is a bit of an apples-to-oranges comparison. The reason that `:kind T` doesn't reduce isn't because the kind of `a` is generalized, but rather because `T` is indexed in `a`, and since `:kind! T` doesn't instantiate `a`, it doesn't have enough arguments to reduce. Note that `:kind! T @Int` _does_ reduce, and this is true regardless of whether the `a` in `T` has its kind generalized or not. (Note that you would observe the same thing if you wrote `type T :: Proxy (a :: Type)`.)

Perhaps you intended to demonstrate an example more along these lines?

```hs
type ProxyType :: Proxy (a :: Type)
type ProxyType = 'Proxy

type T :: Proxy a
type family T where
  T = ProxyType
```

In this example, although the kind of `T` is `forall {k} (a :: k). Proxy a`, `T @Maybe` is stuck, but `T @Int` is not. I think this is the point that you were trying to convey, but double-check me on that.

Reply:
My aim was to show an example where accidental kind-indexing is confusing, but I got confused in the process! I've adapted your example, thanks